ITE: drivers/i2c: I2C driver divided into two compatibles

GTLin08 opened the pull request:
As mentioned in #42882, the I2C of IT8XXX2 is designed for two different
IP blocks, so this PR divides this I2C driver into two compatibles.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
ITE: drivers/i2c: I2C driver divided into two compatibles

As mentioned in #42882, the I2C of IT8XXX2 is designed for two different
IP blocks, so this PR divides this I2C driver into two compatibles.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

keith-zephyr commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,776 @@
+
+/* Set clock frequency for i2c port D, E , or F */
+static void i2c_enhanced_port_set_frequency(const struct device *dev,
+					    int freq_khz)
```
This parameter is now the frequency in Hz, not kHz.

GTLin08 replied:
Done.

keith-zephyr commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,776 @@
+			data->err = ETIMEDOUT;
+			/* reset i2c port */
+			i2c_reset(dev);
+			printk("I2C ch%d:0x%X reset cause %d\n",
```
Use logging instead of printk.

GTLin08 replied:
Done. Also modified line 726.

keith-zephyr commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,776 @@
+	/* Enable I2C function */
+	/* Software reset */
+	IT8XXX2_I2C_DHTR(base) |= IT8XXX2_I2C_SOFT_RST;
+	IT8XXX2_I2C_DHTR(base) &= 0x7F;
```
Use `~IT8XXX2_I2C_SOFT_RST` instead of 0x7F.

GTLin08 replied:
Done.

keith-zephyr commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,776 @@
+		offset = CGC_OFFSET_SMBD;
+		break;
+	case DT_REG_ADDR(DT_NODELABEL(i2c4)):
+		offset = CGC_OFFSET_SMBE;
```
Do I2C channels 4 and 5 still need to update the PMER1 register?

GTLin08 replied:
I2C4 and I2C5 have already set PMER1 register when they are set to FUN3 in pinmux_it8xxx2_set().

keith-zephyr commented on drivers/i2c/i2c_ite_it8xxx2.c:
```diff
@@ -240,15 +179,15 @@ static void i2c_reset(const struct device *dev)
-	IT83XX_SMB_45P3USL = 0x6a;
-	IT83XX_SMB_45P3USH = 0x1;
-	IT83XX_SMB_4P7A4P0H = 0;
+	IT8XXX2_SMB_4P7USL = 0x3;
```
Why are the timing registers changing?  This should be done in a separate PR.

GTLin08 replied:
OK, separated.

keith-zephyr commented on drivers/i2c/i2c_ite_it8xxx2.c:
```diff
@@ -262,61 +201,22 @@ static void i2c_standard_port_set_frequency(const struct device *dev,
 	 * The others use basic 50/100/1000 KHz setting.
 	 */
-	if (freq_khz == 400) {
+	if (freq_khz == I2C_BITRATE_FAST) {
```
The `freq` parameter is now in Hz, not kHz.  Please update.

GTLin08 replied:
Done.

keith-zephyr commented on dts/bindings/i2c/ite,enhance-i2c.yaml:
```diff
@@ -0,0 +1,47 @@
+    reg:
+      required: true
+
+    label:
+      required: true
```
Label is already required by i2c-controller.yaml.  I don't think it needs to be added here.

GTLin08 replied:
Removed.

GTLin08 commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,776 @@
+	 * 1 / (1000 x freq) =
+	 *			2 x (psr + 2) x (1 / (PLL_CLOCK / clk_div))
+	 * psr = ((PLL_CLOCK / clk_div) x
+	 *			(1 / (1000 x freq)) x (1 / 2)) - 2
```
Done.

GTLin08 commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,774 @@
+#include <drivers/pinmux.h>
+#include <errno.h>
+#include <logging/log.h>
+		LOG_MODULE_REGISTER(i2c_ite_enhance);
```
Done.

GTLin08 commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,774 @@
+#include <sys/util.h>
+
+/* Default PLL frequency. */
+#define PLL_CLOCK 48000000
```
We can get the PLL frequency by reading the IT8XXX2_ECPM_PLLFREQR register. Fixed.

GTLin08 commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,774 @@
+		/* Get SMBus clock divide value */
+		clk_div = (IT8XXX2_ECPM_SCDCR2 & 0x0F) + 1U;
+		/* Calculate PSR value */
+		psr = (PLL_CLOCK / (clk_div * (2U * freq_hz))) - 2U;
```
As mentioned above.

GTLin08 commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,774 @@
+	return data->err;
+}
+
+static void enhanced_i2c_start(const struct device *dev)
```
Fixed. Also modified line 517 and 672.

teburd commented on drivers/i2c/i2c_ite_it8xxx2.c:
```diff
@@ -157,53 +125,29 @@ static int i2c_parsing_return_value(const struct device *dev)
-			return -EIO;
-	}
+	/* The device does not respond ACK */
+	if (data->err == HOSTA_NACK)
```
This function has a few missing enclosing brackets around if branch bodies. https://docs.zephyrproject.org/latest/contribute/index.html#coding-style

GTLin08 replied:
Fixed. Thank you.

teburd reviewed (changes requested):
One last code styling issue, looks good otherwise.

keith-zephyr commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,803 @@
+	IT8XXX2_I2C_CTR(base) = E_STS_AND_HW_RST;
+}
+
+static uint32_t i2c_get_pll_frequency(void)
```
This routine is not specific to the I2C module.  Can you move this into [soc.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc%2Friscv%2Friscv-ite%2Fit8xxx2%2Fsoc.c).  You can put the prototype into [soc_common.h](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc%2Friscv%2Friscv-ite%2Fcommon%2Fsoc_common.h)

GTLin08 replied:
Fixed. Thank you.

gmarull commented on drivers/i2c/Kconfig.it8xxx2:
```diff
@@ -8,3 +8,10 @@ config I2C_ITE_IT8XXX2
+
+config I2C_ITE_ENHANCE
+	bool "ITE IT8XXX2 I2C enhance driver"
+	depends on SOC_IT8XXX2
```
please default based on dt_compat_status

GTLin08 replied:
Done.

gmarull commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,771 @@
+#include <drivers/pinmux.h>
+#include <errno.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(i2c_ite_enhance, CONFIG_I2C_LOG_LEVEL);
```
for consistency, can this be put after include list?

GTLin08 replied:
Done.

gmarull commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,771 @@
+ * time and recover bus.
+ */
+struct i2c_alts_cfg {
+	/* Pinmux control group */
+	const struct device *pinctrls;
```
gentle reminder: you should implement a pinctrl driver.

GTLin08 replied:
Pinctrl driver is under development.

gmarull commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,771 @@
+	const struct i2c_enhance_config *config;
+	int res;
+
+	/* Check for NULL pointers */
+	if (dev == NULL) {
+		LOG_ERR("Device handle is NULL");
+		return -EINVAL;
+	}
+	if (msgs == NULL) {
+		LOG_ERR("Device message is NULL");
+		return -EINVAL;
+	}
```
we don't usually make drivers defensive against application programming errors.

GTLin08 replied:
Removed.

gmarull commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,771 @@
+	k_sem_init(&data->device_sync_sem, 0, K_SEM_MAX_LIMIT);
+
+	/* Enable clock to specified peripheral */
+	switch ((uint32_t)base) {
+	case DT_REG_ADDR(DT_NODELABEL(i2c3)):
+		offset = CGC_OFFSET_SMBD;
+		break;
+	case DT_REG_ADDR(DT_NODELABEL(i2c4)):
+		offset = CGC_OFFSET_SMBE;
+		break;
+	case DT_REG_ADDR(DT_NODELABEL(i2c5)):
+		offset = CGC_OFFSET_SMBF;
+		break;
+	}
```
why isn't this stored in each DT instance and retrieved by the driver?

GTLin08 replied:
Done.

gmarull commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,771 @@
+		return error;
+	}
+
+	/* The pin is set to I2C alternate function of SCL */
+	pinmux_pin_set(config->alts_list[SCL].pinctrls,
+		       config->alts_list[SCL].pin,
+		       config->alts_list[SCL].alt_fun);
+	/* The pin is set to I2C alternate function of SDA */
+	pinmux_pin_set(config->alts_list[SDA].pinctrls,
+		       config->alts_list[SDA].pin,
+		       config->alts_list[SDA].alt_fun);
```
ditto, please be aware of https://github.com/zephyrproject-rtos/zephyr/issues/39740

GTLin08 replied:
Ack.

gmarull commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -0,0 +1,771 @@
+	/* Pull SCL and SDA pin to high */
+	gpio_pin_set(config->gpio_dev, config->alts_list[SCL].pin, 1);
+	gpio_pin_set(config->gpio_dev, config->alts_list[SDA].pin, 1);
+	k_msleep(1);
+
+	/* Start condition */
+	gpio_pin_set(config->gpio_dev, config->alts_list[SDA].pin, 0);
+	k_msleep(1);
+	gpio_pin_set(config->gpio_dev, config->alts_list[SCL].pin, 0);
+	k_msleep(1);
+
+	/* 9 cycles of SCL with SDA held high */
+	for (i = 0; i < 9; i++) {
+		/* SDA */
+		gpio_pin_set(config->gpio_dev, config->alts_list[SDA].pin, 1);
+		/* SCL */
+		gpio_pin_set(config->gpio_dev, config->alts_list[SCL].pin, 1);
+		k_msleep(1);
+		/* SCL */
+		gpio_pin_set(config->gpio_dev, config->alts_list[SCL].pin, 0);
+		k_msleep(1);
+	}
+	/* SDA */
+	gpio_pin_set(config->gpio_dev, config->alts_list[SDA].pin, 0);
+	k_msleep(1);
+
+	/* Stop condition */
+	gpio_pin_set(config->gpio_dev, config->alts_list[SCL].pin, 1);
+	k_msleep(1);
+	gpio_pin_set(config->gpio_dev, config->alts_list[SDA].pin, 1);
+	k_msleep(1);
+
+	/* Set GPIO back to I2C alternate function of SCL */
+	pinmux_pin_set(config->alts_list[SCL].pinctrls,
+		       config->alts_list[SCL].pin,
+		       config->alts_list[SCL].alt_fun);
+	/* Set GPIO back to I2C alternate function of SDA */
+	pinmux_pin_set(config->alts_list[SDA].pinctrls,
+		       config->alts_list[SDA].pin,
+		       config->alts_list[SDA].alt_fun);
```
can this procedure be documented? why sleeps of 1ms?

GTLin08 replied:
Add comment.

gmarull commented on soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series:
```diff
@@ -60,6 +60,10 @@ config I2C_ITE_IT8XXX2
 	default y
 	depends on I2C
 
+config I2C_ITE_ENHANCE
+	default y
+	depends on I2C
```
please default based on dt_compat_status so that this can be removed.

GTLin08 replied:
Removed.

gmarull commented on soc/riscv/riscv-ite/it8xxx2/soc.c:
```diff
@@ -51,6 +51,42 @@ static const struct pll_config_t pll_configuration[] = {
 	 .div_usbpd = 5}
 };
 
+uint32_t chip_get_pll_freq(void)
+{
+	uint32_t pllfreq;
+
+	switch (IT8XXX2_ECPM_PLLFREQR & 0x0F) {
+	case 0:
+		pllfreq = MHZ(8);
+		break;
+	case 1:
+		pllfreq = MHZ(16);
+		break;
+	case 2:
+		pllfreq = MHZ(24);
+		break;
+	case 3:
+		pllfreq = MHZ(32);
+		break;
+	case 4:
+		pllfreq = MHZ(48);
+		break;
+	case 5:
+		pllfreq = MHZ(64);
+		break;
+	case 6:
+		pllfreq = MHZ(72);
+		break;
+	case 7:
+		pllfreq = MHZ(96);
+		break;
+	default:
+		return -ERANGE;
+	}
+
+	return pllfreq;
+}
```
why isn't this part of a clock control driver?

GTLin08 replied:
The clock control driver will be implemented later.

GTLin08 committed:
ITE: drivers/i2c: adjust timing registers to pass SI test

Timing registers of I2C 0/1/2 can be adjusted to pass SI test.

We can control the tSU;STA and tHD;DAT simultaneously by changing
the value of the register IT83XX_SMB_4P7USL, and we can control
the tSU;DAT by changing the value of the register IT83XX_SMB_250NS
as well.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

mbolivar-nordic commented on dts/bindings/i2c/ite,it8xxx2-i2c.yaml:
```diff
@@ -33,3 +30,10 @@ properties:
                    The SCL pin must be specified first and the SDA pin
                    second in the pinctrl-0 array.
+
+    clock-gate-offset:
+      type: int
+      required: true
+      description: |
+        The clock gate offsets combine the register offset from
+        ECPM_BASE and the mask within that register into one value.
```
It seems like the following properties are common to both I2C controllers:

- reg
- interrupts
- port-num
- gpio-dev
- pinctrl-0
- clock-gate-offset

Only prescale-scl-low is unique to the 'enhanced' driver. So for maintainability reasons you may want to consider using `include` as is done here for example:

https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/i2c/nordic%2Cnrf-twi-common.yaml
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/i2c/nordic%2Cnrf-twi.yaml
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/i2c/nordic%2Cnrf-twim.yaml

This is not a blocking request, though, just something for you to consider. Thanks for the patch!

GTLin08 replied:
Yes, It can be better maintained, thanks for your advise. Done.

mbolivar-nordic reviewed (dismissed):
Approval is for the DT portions only; I didn't review the rest

GTLin08 commented:
Dear @gmarull, 
Could you review the requested changes, thanks!

GTLin08 commented:
Hi @teburd, @keith-zephyr, @mbolivar-nordic 
I need to get your approval again, thanks!

GTLin08 commented:
Dear sirs,

Truly grateful for helps. However, we’d like to invite you back to review the PR #42812 which is about the I2C 50KHz option. Thanks again for your great patience and guidance.
